torch/csrc/autograd/functions/accumulate_grad.cpp

```diff
@@ -64,9 +64,9 @@ auto AccumulateGrad::apply(variable_list&& grads) -> variable_list {
     // a thing never promised and documented, but used in some hacks seen
     // on the internet.
     if (grad_variable.is_sparse() && !new_grad.is_sparse()) {
-      grad_variable.set_data(new_grad.data() + grad_variable.data());
+      grad_variable = new_grad + grad_variable;
```

Comment:
In the old `variable.set_data(data_tensor)` method, it was ok to pass a sparse `data_tensor` to a dense `variable` (or vice versa). In the new `variable.set_data(data_tensor)` method, this behavior is not possible to implement (we can't just swap out the variable's TensorImpl and put in a SparseTensorImpl), we instead choose to disallow this behavior, and fix places where this behavior is expected (which only occurs here and in `test_optim.py`).

Reply:
So... at the end of the day, this mutable reference is into `variable.grad();`? I guess... that works

Reply:
Another note: this does not accumulate gradient, because it is in the `!GradMode::is_enabled()` if-branch.

Reply:
I'm not sure if this is still upholding the "more predictable" goal -- can you give an example where we'd hit the previous code in python?

Reply:
We would hit the previous code in the `TestOptim.test_adagrad_sparse` test case, in the following way:
```python
def _test_rosenbrock_sparse(self, constructor, scheduler_constructors=None,
                            sparse_only=False):
    if scheduler_constructors is None:
        scheduler_constructors = []
    params_t = torch.Tensor([1.5, 1.5])

    params = Variable(params_t, requires_grad=True)
    optimizer = constructor([params])
    schedulers = []
    for scheduler_constructor in scheduler_constructors:
        schedulers.append(scheduler_constructor(optimizer))

    if not sparse_only:
        params_c = Variable(params_t.clone(), requires_grad=True)
        optimizer_c = constructor([params_c])

    solution = torch.Tensor([1, 1])
    initial_dist = params.data.dist(solution)

    def eval(params, sparse_grad, w):
        # Depending on w, provide only the x or y gradient
        optimizer.zero_grad()
        loss = rosenbrock(params)
        loss.backward()
        grad = drosenbrock(params.data)
        # NB: We torture test the optimizer by returning an
        # uncoalesced sparse tensor
        if w:
            i = torch.LongTensor([[0, 0]])
            x = grad[0]
            v = torch.DoubleTensor([x / 4., x - x / 4.])
        else:
            i = torch.LongTensor([[1, 1]])
            y = grad[1]
            v = torch.DoubleTensor([y - y / 4., y / 4.])
        x = sparse.DoubleTensor(i, v, torch.Size([2]))
        with torch.no_grad():
            if sparse_grad:
                params.grad = x
            else:
                params.grad = x.to_dense()
        return loss

    for i in range(2000):
        # Do cyclic coordinate descent
        w = i % 2
        optimizer.step(functools.partial(eval, params, True, w))
        for scheduler in schedulers:
            if isinstance(scheduler, ReduceLROnPlateau):
                scheduler.step(rosenbrock(params))
            else:
                scheduler.step()
        if not sparse_only:
            optimizer_c.step(functools.partial(eval, params_c, False, w))
            self.assertEqual(params.data, params_c.data)

    self.assertLessEqual(params.data.dist(solution), initial_dist)
```
In this test example, the first iteration of `for i in range(2000):` will run `params.grad = x` where `x` is a sparse tensor, and after that step `params.grad` is a sparse tensor. In the second iteration of `for i in range(2000):`, at the `loss.backward()` step, in `AccumulateGrad::apply` the `if (grad_variable.is_sparse() && !new_grad.is_sparse()) {` branch will be called, because we now have a sparse `grad_variable` and a dense `new_grad`, and the original code `grad_variable.set_data(new_grad.data() + grad_variable.data())` was expected to change `grad_variable` from a sparse tensor to a dense tensor (i.e. the result of `new_grad.data() + grad_variable.data()`).

The issue here is that using `grad_variable.set_data(new_grad + grad_variable)` to change `grad_variable` from a sparse tensor to a dense tensor is not possible anymore, and as a result the mentioned goal of "avoid changing the grad tensor" is not achievable anymore with the new `set_data()` semantics.

Reply:
An example that would hit the previous code looks like this:
```python
params = torch.Tensor([1.5, 1.5]).requires_grad_()
y = torch.FloatTensor([1.]).requires_grad_()[0]
with torch.no_grad():
    # Change gradient to a sparse tensor
    params.grad = torch.sparse.FloatTensor(torch.LongTensor([[1, 1]]), torch.FloatTensor([y, y]), torch.Size([2]))

loss = (params[1] - params[0] ** 2)
grad_saved = params.grad
loss.backward()  # This will hit the `grad_variable.is_sparse() && !new_grad.is_sparse()` branch
assert id(grad_saved) == id(params.grad)
# The above assert doesn't fail in current master, but would fail if we call `grad_variable = new_grad + grad_variable` in the if-branch
```
This example is a simplification of part of the logic in `test_adagrad_sparse`/`test_sgd_sparse`/`test_sparse_adam` in test_optim.py, and the `grad_variable.is_sparse() && !new_grad.is_sparse()` branch is also only hit in those three tests. 

I think there are two options we can take here:
1. Don't uphold the goal of "[not] changing the grad tensor" in this if-branch (i.e. let `assert id(grad_saved) == id(params.grad)` fail), because in those tests we have to change `params.grad` from a sparse tensor to a dense tensor, and that operation has to change the tensor itself.
2. Implement an `._internal_set_impl()` API that works the same as the original `.data = ` API and allows a user to change a tensor's TensorImpl type (e.g. from sparse to dense). This can uphold the invariant of `assert id(grad_saved) == id(params.grad)`, but it would only be useful for `test_adagrad_sparse`/`test_sgd_sparse`/`test_sparse_adam` in test_optim.py.

Reply:
one way to make this work would probably be copying over the pyobj poitner, but that seems too hacky...

Reply:
@gchanan I can confirm that copying over the pyobj pointer works:
```cpp
PyObject* pyobj_saved = grad_variable.unsafeGetTensorImpl()->pyobj();
grad_variable = new_grad + grad_variable;
grad_variable.unsafeGetTensorImpl()->set_pyobj(pyobj_saved);
```
I feel that it's okay to copy over the pyobj pointer, because it makes the PR less BC-breaking. As long as the hack is used only in this case, it should still be manageable.

Reply:
Update from in-person discussion: the use case mentioned in https://github.com/pytorch/pytorch/pull/17072/files#r285834118 seems rare, so we decide to not uphold the goal of "[not] changing the grad tensor" in this if-branch. We will revisit this decision if in the future there is a legitimate reason why we need to keep the grad tensor unchanged.